### PR TITLE
[OpenGL] use GLAD as API loader (replace GLEW)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #664 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
`glad` is not a library, it's a code generator, a python package.
We can generate our own `external/glad`, without any library loading required.

**Comment by archibate** [comment]
Local test passing.

File: taichi/backends/opengl/opengl_api.cpp [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #include "taichi/program/program.h"
 
 #ifdef TI_WITH_OPENGL
-#include "GL/glew.h"
+#include <glad/glad.h>
```
Comment (yuanming-hu) [inline taichi/backends/opengl/opengl_api.cpp:9]: Let's use `#include "glad/glad.h"` here.

**Comment by yuanming-hu** [comment]
> Local test passing. Do you have a win environment?

I do. I can help test on Windows once this is merged.

**Comment by archibate** [comment]
> Thanks! Let's see if `glad` is more portable than `glew` (I believe so). Only major thing to fix here: the `glad` files are too big (`1.6 MB`) in total - could you please use a submodule (just like https://github.com/taichi-dev/glew-ready)?

To be clear this is not a repo, glad is a command line tool that generates this dir (`glad --out-path=external/glad`). So two choices:
1. taichi-dev/glad-ready as submodule
2. let user run glad themselves (may harm gfw users since it downloads a huge xml file from khr's offical repo)

1 sounds far more better but I'm not sure if the `external/glad` generated in Linux can work on Windows (and other Linux dist) too, if you could do me a favor...

**Comment by yuanming-hu** [comment]
> To be clear this is not a repo, glad is a command line tool that generates this dir (`glad --out-path=external/glad`).

I know, we should probably just include the **generated files instead of the glad source** in the submodule, and update the submodule when necessary.

You might need to generate both GLX (Linux) and WGL (Windows) versions using https://glad.dav1d.de/ or a downloaded glad.

I'm happy to test for you on Windows when you are ready.

**Comment by archibate** [comment]
> You might need to generate both GLX (Linux) and WGL (Windows) versions using https://glad.dav1d.de/ or a downloaded glad.

It seems like current glad.c already includes WGL support:

https://github.com/taichi-dev/glad-ready/blob/6e0265ca71db319da02fe4bdfead2430e089d510/src/glad.c#L670

It detects `_MSC_VER` and `_WIN32`, seems it's generic if no `--spec` specified.

**Comment by yuanming-hu** [comment]
Now it works on Windows :-) Great job!

image

**Comment by archibate** [comment]
All done! Except for:
> Hide/remove https://github.com/taichi-dev/glew-ready from taichi-dev

Sadly I can't find the `Settings` section in this repo... maybe I don't have admin access to it? Could you remove it or transfer ownership to me? Thanks!

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -256,11 +252,14 @@ def supported_archs():
     wanted_archs = os.environ.get('TI_WANTED_ARCHS', '').split(',')
     # Note, ''.split(',') gives you [''], which is not an empty array.
     wanted_archs = list(filter(lambda x: x != '', wanted_archs))
-    if len(wanted_archs):
+    excluded_archs = list(filter(lambda x: x.startswith('^'), wanted_archs))
```
Comment (yuanming-hu) [inline python/taichi/lang/__init__.py:255]: Please, revert this part, do this in a different PR and use `ti test -a x64` for now. The correctness of this part is not trivial and thereby needs discussions. In this PR I'd rather focus on GLAD.

**Comment by yuanming-hu** [comment]
> All done! Except for:
> 
> > Hide/remove https://github.com/taichi-dev/glew-ready from taichi-dev
> 
> Sadly I can't find the `Settings` section in this repo... maybe I don't have admin access to it? Could you remove it or transfer ownership to me? Thanks!

Thanks for fixing everything. Now you are the admin of `glew-ready`. Could you try again?

**Comment by archibate** [comment]
CI is passed!
